hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/WatermarkCoalescer.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.execution;
+
+import com.hazelcast.jet.JetException;
+import com.hazelcast.jet.core.Watermark;
+import com.hazelcast.jet.impl.util.TimestampHistory;
+
+import java.util.Arrays;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+/**
+ * Implements {@link Watermark} coalescing. Tracks WMs on queues and decides
+ * when to forward the WM. The watermark should be forwarded:
+ * <ul>
+ *     <li>when it has been received from all input streams
+ *     <li>if the maximum watermark retention time has elapsed
+ * </ul>
+ */
+class WatermarkCoalescer {
+
+    private final long[] queueWms;
+    private final TimestampHistory watermarkHistory;
+
+    private long lastEmittedWm = Long.MIN_VALUE;
+    private long topObservedWm = Long.MIN_VALUE;
+
+    WatermarkCoalescer(int maxWatermarkRetainMillis, int queueCount) {
+        queueWms = new long[queueCount];
+        Arrays.fill(queueWms, Long.MIN_VALUE);
+
+        watermarkHistory = maxWatermarkRetainMillis >= 0
```

Comment:
I think we could skip holding any history if queue count == 1, as it's equivalent to not having history at all. This will prevent checking time continuously in processors / edges with only upstream queue.

It might be easier to use a specialized class for this case which will be a lot simpler.

Reply:
fixed